packages/webamp/js/actionCreators/files.ts

```diff
@@ -400,3 +401,62 @@ export function addDirAtIndex(nextIndex: number): Thunk {
     );
   };
 }
+
+export function addFilesFromUrl(atIndex = 0): Thunk {
+  return async (dispatch, getState, { handleAddUrlEvent }) => {
+    if (handleAddUrlEvent) {
+      const tracks = await handleAddUrlEvent();
+
+      if (tracks != null) {
+        dispatch(loadMediaFiles(tracks, LOAD_STYLE.NONE, atIndex));
+        return;
+      }
+    } else {
+      alert("Not supported in Webamp");
+    }
+  };
+}
+
+export function addFilesFromList(): Thunk {
+  return async (dispatch, getState, { handleLoadListEvent }) => {
+    if (handleLoadListEvent) {
+      const tracks = await handleLoadListEvent();
+
+      if (tracks != null) {
+        dispatch(removeAllTracks());
+
+        dispatch(loadMediaFiles(tracks, LOAD_STYLE.NONE, 0));
+        return;
+      }
+    } else {
+      alert("Not supported in Webamp");
+    }
+  };
+}
+
+export function saveFilesToList(): Thunk {
+  return (dispatch, getState, { handleSaveListEvent }) => {
+    if (handleSaveListEvent) {
+      const state = getState();
+      const trackIds = getTrackOrder(state);
+      const tracks = getTracks(state);
+      handleSaveListEvent(
+        trackIds.map((id) => {
+          const track = tracks[id];
+
+          return {
+            url: track.url,
+            metaData: {
+              artist: track.artist || "",
+              title: track.title || "",
+              album: track.album,
+              albumArtUrl: track.albumArtUrl || "",
+            },
+          };
```

Comment:
For bonus points we could move the generation of this array of tracks into a selector in the selectors file. (Not required)

Reply:
Sure, I name it getUserTracks since Track an object type to interact with the playlist user and also because getTracks was already taked.